Turn the tests back on and fix everything that it breaks

UniversalSuperBox opened the pull request:
1.5 months later...

These commits turn on the Unity8 qmltests and run them automatically during a deb build using dh_auto_test. Then, we fix the myriad race conditions caused by not checking if a component was actually in a state to receive input, or not waiting for the effects of that input to resolve.

This isn't all a bunch of busywork though, it gave me insight into fixing the following bugs with window management:

* Fixes https://github.com/ubports/ubuntu-touch/issues/1081
* Fixes https://github.com/ubports/ubuntu-touch/issues/1124

It will alert us in the future if we do something bad and break some tested behavior. We will start adding tests to confirm and fix bugs in the same PR we fix them.

This does crank up the time to build in CI to an hour. The only way we'll fix this is by optimizing Unity8 itself.

If anyone ever turns the tests off again, they're getting whacked over the head with a Nexus 7.

mariogrip committed:
Merge pull request #68 from vanyasem/bionic

Fix Qt >= 5.9

UniversalSuperBox committed:
Use the xcb QPA when running xvfbtests

UniversalSuperBox committed:
Appropriately auto-test

UniversalSuperBox committed:
Fix copyright tests

UniversalSuperBox committed:
Do not install scopefakes

UniversalSuperBox committed:
Use position rather than ID to determine touch point

Using the id() of a TouchPoint does not give the same ID used to create
and move it. The ID of .press() and friends only determine the finger
or cursor to act on. This may have worked before due to an
implementation detail of the xcb or Mir QPAs.

See also https://doc.qt.io/qt-5/qtouchevent-touchpoint.html#id

UniversalSuperBox committed:
Remove test to check if Dash remains running

The Dash has been removed, this test is not needed.

dobey committed:
Run tests in parallel and sync outputs.

This runs the build (and tests) with -O option to sync output from make for
each target, to make it more readable. This also runs all the tests with -k
option so make will keep going and run other tests in the event one fails.

dobey committed:
Need the Ubuntu.Test module too.

dobey committed:
Need system-settings for wizard tests.

dobey committed:
Remove unused Carousel component.

dobey committed:
Need Ubuntu.Layouts for qml tests too.

UniversalSuperBox committed:
Skip testIconLookupLogic, it requires local config

UniversalSuperBox committed:
Make right edge tutorial only take 3 apps to open

In older versions of Unity8, the Dash was open at all times, counting as
an Application for ApplicationManager. Now that the Dash is no longer,
reduce the needed application count to three to retain the same user
experience.

UniversalSuperBox committed:
Always hide the right tutorial when the Spread shows

Ignoring the spread showing was likely done so that keyboard+phone users
couldn't dismiss the Tutorial without using the touch screen.
This is no longer a requirement and it sometimes lets you open the
Spread without dismissing the tutorial... after which you can't dismiss
the tutorial any more.

UniversalSuperBox committed:
Skip the rightShortDrag tutorial test

Either test_tutorialRightShortDrag or test_tutorialRightAutoSkipped may
run in a single session under 16.04's standard xvfb, not both. Give up
in ever finding the problem by skipping test_tutorialRightShortDrag

UniversalSuperBox committed:
Fix MenuBar TestCase failures

* Set modelData to null on creation to ensure the first test gets init
* Move the mouse to (0,0) before starting the test... The mouse not
  getting reset caused all sorts of errors
* Clear aboutToShowCalledSpy
* In test_disabledTopLevel and test_firstDisabled, wait for events to
  finish to avoid racing
* Add an XFAIL test to alert to incorrect behavior when the mouse hangs
  over the first menu item it clicks

UniversalSuperBox committed:
Reduce MenuItem warnings

I'm not sure why parent isn't set on RowLayout, but it's able to anchor
correctly and doesn't complain when referencing root.

UniversalSuperBox committed:
Wizard: Update tests, remove app updates

The implementation of app updates in the Wizard is messy, difficult to
test, and doesn't need to be carried any more. Because of that, I've
removed the app update feature.

I've also:

* Added mock for SystemImage
* Updated the System and NetworkingStatus mocks to match their upstreams
* Moved around setup and teardown functions for each test
* Removed tests for the location pages
* Added a test for the system update page

UniversalSuperBox committed:
Correct tryCompare syntax

UniversalSuperBox committed:
Wait for the click to get registered by actionSpy

UniversalSuperBox committed:
Fix the tests on the Panel

* Put the mouse somewhere neutral so it doesn't hover over stuff
* Staged and Windowed modes were switched in design, correct that.
* Be more careful about trusting things to not be null / wait for
  rendering more often

UniversalSuperBox committed:
Refactor the Drawer for easier testing

Move most of the behavior for the Drawer out of Launcher and into
Drawer, using signals to acheive anything that Drawer can't do by itself

UniversalSuperBox committed:
Update/Fix Drawer tests

This was going to be the interesting case since I'd recently made
changes to the Drawer.

* Make resetOldFocus reset focus and our knowledge of it
* give appList an objectName, change tests to use it
* DRY with openDrawerAndWait
* Verify keyboard navigation correctly

UniversalSuperBox committed:
Fix up the Launcher tests

After a mouseClick, the mouse now hovers at the click point. Switch the
order of movements so the mouse actually moves.

UniversalSuperBox committed:
Fix up shell tests

* Use initTestCase for first-test setup
* Move the mouse to a neutral position before running a test
* Reorder functions so all tests are contiguous
* Turn on an accidentally disabled test (test_highlightFollowsMouse) and
  update it for the newest changes
* Wait for longer and use tryCompare more often to allow asynchronous
  events to fire

UniversalSuperBox committed:
More Shell test fixing

* Further defragment tests and helper functions
* When moving the mouse and making sure you didn't go too far on the x
  axis, checking your y axis is not very useful.

UniversalSuperBox committed:
Fix still-flakey Shell tests

* Spin loops slower so they can process
* Use mouseFlick rather than the loop in progressiveAutoScrolling

UniversalSuperBox committed:
Loosely compare state value with the Mir enum

For whatever reason, value (or model.window.state) are not the same as
the Mir::State enum we're comparing against... at least on armhf and
arm64. This means that none of the functionality to visibly change state
was being called. By loosely comparing we get the correct behavior.

UniversalSuperBox committed:
Fix racing in MenuPopup tests

* Move the mouse out of the way at the start of tests
* Fix use of compare vs tryCompare, which caused a race condition

UniversalSuperBox committed:
Accept asynchronicity in more MenuBar tests

UniversalSuperBox committed:
Fix handling of fullscreen on ARM

* Loosely compare against the Mir::State enum everywhere since it
  doesn't play nice with strict comparison
* Adjust the ApplicationInfo mock so it does not return fullscreen() to
  QML. This ensures that Unity is using the application surface's state
  rather than falling back to the application's state. Better for multi
  window apps.
* In test_appRotatesWindowContents, use the Surface rather than the
  Application when appropriate

UniversalSuperBox committed:
Fix up the TabletStage tests

* Do not launch the Dash, it doesn't exist any more.
* Add buttons to let users of "try" move apps
* Reduce various race conditions
* Add a launchApp function to reduce copied code
* Remove the test for the Dash being dragged about

UniversalSuperBox committed:
Don't run tests in parallel :/

UniversalSuperBox committed:
Better check for the Drawer's readiness

The Drawer tests were able to continue running before they had fully
extended the Drawer. Use the Drawer's fullyOpen property to make sure
that doesn't happen.

UniversalSuperBox committed:
Launcher tests: Convert any async to try*

To greatly reduce the chance of tests flapping between pass or fail,
convert any checks that are based on asynchronous events (like mouse
clicks) to tryCompare or tryVerify.

UniversalSuperBox committed:
Squash race conditions everywhere

I searched the entire codebase for compare() entries and evaluated all
of them for how appropriate they are compared to tryCompare() or another
alternative. I found that there were many cases where compare()
depended on delays that came before it to get the correct value, causing
race conditions that we haven't even hit yet. In a proactive measure,
this commit:

* Replaces compare() with tryCompare() where appropriate
* Replaces tryCompareFunction() with tryVerify() where appropriate
* Uses SignalSpy.wait() where appropriate
* Generally lints and pays back technical debt when I found the
  opportunity

UniversalSuperBox committed:
SwipeToAct test: Ensure slider is ready to slide

UniversalSuperBox committed:
Lockscreen test: Clean up test_resize, tryVerify

UniversalSuperBox committed:
Launcher tests: Ensure the quickList is ready

dobey reviewed (approved):
Yay tests!

mardy commented on debian/control:
```diff
@@ -72,6 +74,7 @@ Build-Depends: cmake,
                qtdeclarative5-ubuntu-content1,
                qtdeclarative5-ubuntu-settings-components (>= 0.11),
                ttf-ubuntu-font-family,
+               ubuntu-system-settings,
```
Why is this needed? Are you sure you don't need `libsystemsettings-dev` instead?

UniversalSuperBox replied:
It was needed for the Wizard tests for app updates, which were then removed by 6653caea530347286883bea94f8d3ef4595aa91c. I'll remove it here.

mardy commented on qml/ApplicationMenus/MenuItem.qml:
```diff
@@ -77,11 +77,11 @@ ActionItem {
         anchors.leftMargin: units.gu(1)
         anchors.rightMargin: units.gu(1)
-        anchors.verticalCenter: parent.verticalCenter
+        anchors.verticalCenter: root.verticalCenter
```
This change (and the two above it) are not needed, and when using anchors it's preferable to use `parent`, than referencing by id. I would revert these.

UniversalSuperBox replied:
Trying to use `parent` caused QML warnings and the items didn't anchor correctly. See 468a1a62144063816dd355a6a715e9fc056a5f81

mardy replied:
what was the warning? Does it go away if you add a binding `parent: root`?

UniversalSuperBox replied:
```
QWARN  : qmltestrunner::MenuBar::test_stray_submenus() file:///home/dalton/unity8/qml/ApplicationMenus/MenuItem.qml:80: TypeError: Cannot read property of null
QWARN  : qmltestrunner::MenuBar::test_stray_submenus() file:///home/dalton/unity8/qml/ApplicationMenus/MenuItem.qml:81: TypeError: Cannot read property of null
QWARN  : qmltestrunner::MenuBar::test_stray_submenus() file:///home/dalton/unity8/qml/ApplicationMenus/MenuItem.qml:84: TypeError: Cannot read property of null
```

I can log parent in `Component.onCompleted` and get `qml: MenuItem_QMLTYPE_12_QML_18(0x1d43d50, "menuBar-item0-menu-item1-actionItem")` as expected.

This does not go away with a `parent: root` binding.

mardy replied:
OK for your changes, but I wonder if we should file a bug on the uitk's ActionItem.

mardy commented on qml/Launcher/Drawer.qml:
```diff
@@ -29,12 +29,85 @@ FocusScope {
+
+    // Request that the Drawer (and maybe its parent) is hidden, normally if
+    // the Drawer has been dragged away.
+    signal hide();
```
I would call these `openRequested` and `hideRequested`, so that on the receiving side you'll have `onOpenRequested` and `onHideRequested`. `open()` and `hide()` are good names for functions, not for signals (see also `applicationSelected` above: it's not just `select`).

mardy replied:
Also, remove the `;` at the end of those lines

mardy commented on qml/Launcher/Drawer.qml:
```diff
@@ -29,12 +29,85 @@ FocusScope {
+                // See onDraggingHorizontallyChanged below
+                searchField.focus = hadFocus;
+                searchField.select(oldSelectionStart, oldSelectionEnd);
+                return;
```
you don't need `return`

mardy commented on qml/Shell.qml:
```diff
@@ -552,7 +552,7 @@ StyledItem {
 
             readonly property bool focusedSurfaceIsFullscreen: topLevelSurfaceList.focusedWindow
-                ? topLevelSurfaceList.focusedWindow.state === Mir.FullscreenState
+                ? topLevelSurfaceList.focusedWindow.state == Mir.FullscreenState
```
This looks very suspicious; if it works with the new code but not with the old, it looks like we might have some error in registering the enums.

UniversalSuperBox replied:
I agree there's no reason, but the bug is architecture-specific. Strict comparison worked on amd64 but not armhf or arm64. The values being returned are correct but for whatever reason they don't have the right type.

mardy replied:
Could you please file a bug about this, so when someone has some time he can investigate this issue?

mardy commented on qml/Stage/Stage.qml:
```diff
@@ -967,34 +967,34 @@ FocusScope {
                     }
                     onStateChanged: {
-                        if (model.window.state === Mir.MinimizedState) {
+                        if (value == Mir.MinimizedState) {
```
Same here: there's no reason why the old code wouldn't work, unless we have some serious problem.

mardy commented on qml/Stage/TabletStage.qml:
```diff
@@ -775,7 +775,7 @@ AbstractStage {
                             return priv.mainStageDelegate.fullscreen;
                         } else if (surface) {
-                            return surface.state === Mir.FullscreenState;
+                            return surface.state == Mir.FullscreenState;
```
same here

mardy commented on qml/Stage/WindowedFullscreenPolicy.qml:
```diff
@@ -23,7 +23,7 @@ import Unity.Application 0.1
 QtObject {
     function applyPolicy(surfaceState, surfaceChrome) {
-        if (surfaceState === Mir.FullscreenState && surfaceChrome === Mir.LowChrome) {
+        if (surfaceState == Mir.FullscreenState && surfaceChrome === Mir.LowChrome) {
```
same here

mardy reviewed (changes requested):
Nice to see the tests coming back. I've left a few inline comments; nothing super critical, though.

UniversalSuperBox committed:
Fix up Drawer based on code review

* Give signals more descriptive names
* Don't return uselessly in onRightMarginChanged
* Remove semicolons where they aren't needed

UniversalSuperBox committed:
ubuntu-system-settings is not needed for tests

After 6653caea530347286883bea94f8d3ef4595aa91c, we no longer need
ubuntu-system-settings to test the Wizard.
Also remove the commented-out entry since it isn't needed.

UniversalSuperBox commented:
I've fixed some of these issues, @mardy. Could I get another review?

mardy reviewed (approved):
Thanks! This looks good to me; I'm still puzzled by the substitution of `===` to `==` when comparing enums; maybe you could file a bug about it, in order to document the problem (and hopefully address it, once someone gets some time to look into it)?

UniversalSuperBox commented:
https://github.com/ubports/unity8/issues/148
